tests/cxx/typedef_in_template.cc

```diff
@@ -7,51 +7,54 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include "tests/cxx/typedef_in_template-d1.h"
-
-template<class T>
-class Container {
- public:
-  // Should not be an iwyu violation for T
-  typedef T value_type;
-
-  // C++11 alias declaration, should not be an iwyu violation for T
-  using alias_type = T;
-
-  // IWYU: Pair is...*typedef_in_template-i1.h
-  typedef Pair<T,T> pair_type;
+#include "direct.h"
+
+// Verify that a full-use of an alias of a template parameter is treated as a
+// full-use of that parameter.
+template<typename T>
+struct A {
+  using TAlias = T;
+  TAlias t;
 };
 
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
+A<IndirectClass> a;
 
-void Declarations() {
-  // These do not need the full type for Class because they're template params.
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
+A<IndirectClass>::TAlias a2;
 
-  // TODO: This is almost certainly wrong, see bug #431
-  // We should not require the full definition of Class for passing it as a
-  // template argument, but we must require it when the typedef it's aliasing
-  // is full-used.
-  // The bug has instructions for how to provoke the error more obviously.
-
-  // IWYU: Class needs a declaration
-  Container<Class>::value_type vt;
```

Comment:
Same here, we need to make sure that `value_type` requires `Container`, but does not force `Container` to require `Class`. I think your new testcases demonstrate this, but it's not as easily mappable back to STL-like types.

Reply:
I've restored the deleted test cases.